docs/content/docs/concepts/Breaking Changes and Supported Patterns.md

```diff
@@ -0,0 +1,293 @@
+However adding the optional `override` causes the output array to be `[1, 2, 3]` which would almost certainly violate the documented behavior from before the extra argument was added. This can be avoided by doing any of the following:
+
+1. When passing a function in a way that could implicitly convert its type, wrap it in a lambda (in the example, replace `processItem` at the end with `(s) => processItem(s)`).
+2. Avoid casting functions to have more arguments. In the example replace `(item: string) => string)` with `(item: string, index: number) => string)`
+3. Don't have optional arguments.
+4. Pick one of the above for every function type usage.
+
+The first two approaches are hard to audit/enforce but allows the package to add optional arguments to functions.
+They also make adding extra arguments to callback function types an incompatible change.
+
+This makes the most pragmatic solution to do all of:
+
+1. Try and do #2 above (avoid casting functions to have more arguments). The package must do this everywhere it's relevant and technically only User Code that follows this will be considered a supported pattern.
+2. Since we are relying on avoiding implicit casting away of arguments for correctness, adding new arguments to callback functions needs to be considered an Incompatible change, even though it almost always compiles and usually works.
+3. Prefer patterns that avoid optional arguments since enforcing these rules is hard and unlikely to be done robustly without compiler assistance.
+
+Some alternatives to using optional arguments:
+
+1. The "keyword arguments" pattern, where an object is provided with named optional fields instead of optional arguments.
+   This also solves the issue of order dependency of optional arguments and is great when several optional arguments are needed.
+   The type of the object used can be an interface with the "in" variance as described in "Version Type Variance" above to allow adding members.
+2. Add alternative functions that take the extra arguments.
+   Can be combined with the above to future proof it for when even more arguments are added.
+3. Make the existing arguments more flexible.
+   For example if the function takes one string, you can replace that argument with a string or configuration object that contains the string and extra optional arguments.
+
+## Module name collisions
+
+User Code which could be broken by a member being added to a module in the Package should not be supported.
+This specifically means User Code that does things like:
+
+```typescript
+export * from "the-package";
+export { nameThatCouldCollide };
+```
+
+Is not supported.
+
+## Imports for other than the root
+
+User code which imports any path not at the root of the package is not supported.
+This allows refactoring the package's internals.
+
+## Undocumented behavior
+
+User code which depends on any behavior of the package which is undocumented should be unsupported.
```

Comment:
I like this as a rule generally, but it does mean we need to be extra diligent in documenting our APIs thoroughly. I don't think it's fair to our customers, as a somewhat extreme example, to export something entirely undocumented and not treat forwards and backwards compatibility with case.